File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        answer = self.cal(root)

        return max(answer)

    def cal(self, root):
        if root.left and root.right:
            l = self.cal(root.left)
            r = self.cal(root.right)
            return [max(l[0], r[0]) + 1, max(l[0] + r[0] + 2, l[1], r[1])]
        if root.left and not root.right:
            l = self.cal(root.left)
            return [l[0] + 1, l[1]]
        if not root.left and root.right:
            r = self.cal(root.right)
            return [r[0] + 1, r[1]]
        if not root.left and not root.right:
            return [0, 0]

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  int diameterOfBinaryTree(TreeNode? root) {
    var answer = cal(root);
    return max(answer[0],answer[1]);
  }
  List<int> cal(TreeNode? root) {
    if (root?.right == null && root?.left == null) {
        return [0, 0];
    } else if (root?.right == null && root?.left != null) {
        var left_list = cal(root?.left);
        return [left_list[0] + 1, max(left_list[0] + 1, left_list[1])];
    } else if (root?.left == null && root?.right != null) {
        var right_list = cal(root?.right);
        return [right_list[0] + 1, max(right_list[0] + 1, right_list[1])];
    } else {
        var left_list = cal(root?.left);
        var right_list = cal(root?.right);

        return [max(left_list[0]+1,right_list[0]+1),
        max(left_list[0]+1+right_list[0]+1, max(right_list[1],left_list[1]))
        ];
    }

  }
}